applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -919,6 +929,116 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::ComputePotentia
+
+    // make normals point into element
+    first_edge_normal[0] = -1.0 * first_edge_normal[0];
+    first_edge_normal[1] = -1.0 * first_edge_normal[1];
+
+    second_edge_normal[0] = -1.0 * second_edge_normal[0];
+    second_edge_normal[1] = -1.0 * second_edge_normal[1];
+
+    third_edge_normal[0] = -1.0 * third_edge_normal[0];
+    third_edge_normal[1] = -1.0 * third_edge_normal[1];
+
+    // get free stream velocity and norm
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    const double free_stream_velocity_norm = MathUtils<double>::Norm3(free_stream_velocity);
+
+    // find component of element edge normal vector in direction of free stream velocity
+    const double first_edge_normal_fs_comp = inner_prod(first_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+    const double second_edge_normal_fs_comp = inner_prod(second_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+    const double third_edge_normal_fs_comp = inner_prod(third_edge_normal, free_stream_velocity) / free_stream_velocity_norm;
+
+    vector<double> edge_normal_velocity_components (3);
+    edge_normal_velocity_components[0] = first_edge_normal_fs_comp;
+    edge_normal_velocity_components[1] = second_edge_normal_fs_comp;
+    edge_normal_velocity_components[2] = third_edge_normal_fs_comp;
+
+    this->SetValue(EDGE_NORMAL_VELOCITY_COMPONENTS, edge_normal_velocity_components);
+
+    // get node IDs of upwind element
+    int upwind_element_one_node_id = 0;
+    int upwind_element_two_node_id = 0;
+
+    // vector between points 0 and 1 has biggest streamwise component
+    if(first_edge_normal_fs_comp > second_edge_normal_fs_comp && first_edge_normal_fs_comp > third_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 0;
+        upwind_element_two_node_id = 1;
+    }
+    // vector between points 1 and 2 has biggest streamwise component
+    else if(second_edge_normal_fs_comp > first_edge_normal_fs_comp && second_edge_normal_fs_comp > third_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 1;
+        upwind_element_two_node_id = 2;
+    }
+    // vector between points 0 and 2 has biggest streamwise component
+    else if(third_edge_normal_fs_comp > second_edge_normal_fs_comp && third_edge_normal_fs_comp > first_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 0;
+        upwind_element_two_node_id = 2;
+    }
+    else
+    {
+         KRATOS_WARNING("Transonic perturbation element: FindUpWindElement") <<
+        "no maximum streamwise component found" << std::endl;
+    }
+
+    // find elements attached to each node
+    const GlobalPointersVector<Element>& rNodeOneElementCandidates = r_geom[upwind_element_one_node_id].GetValue(NEIGHBOUR_ELEMENTS);
+    const GlobalPointersVector<Element>& rNodeTwoElementCandidates = r_geom[upwind_element_two_node_id].GetValue(NEIGHBOUR_ELEMENTS);
```

Comment:
What you do here is correct :+1: Could you try to think how to generalize it so that it works both in 2d and 3d? One way would be to do:

https://github.com/KratosMultiphysics/Kratos/blob/5f498423d3c724ddba5ca5ed7a4abcc246c563ef/applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp#L289-L299